### DISCUSSION TITLE
DRF's  serializer's  SerializerMethodField  are shown in  .get_fields()

### DISCUSSION THREAD

**Discussion by Teut2711** [encode/django-rest-framework#9537] [Ideas & Suggestions]
Hello everyone,

I wanted to check if the current behavior I'm observing is expected.

Currently, calling .get_fields() on an instance of a serializer returns all fields, including computed fields such as SerializerMethodField. In my use case, I'm parsing a complex text file and using serializers for validation, among other things. After parsing, I end up with a list of CSV rows, which I feed one by one into the serializer.

My expectation is that I should be able to create a dictionary from .get_fields().keys() and the row containing the data. Ideally, only the non-computed fields should be returned. After mapping the data and passing it to the serializer for validation, the computed fields would then be generated.

Is this behavior intentional, or is there a recommended approach to exclude computed fields when calling .get_fields()?

**Comment by lovelydinosaur** [comment]
For your use case I think you want to either inspect read_only or inspect the type.

**Reply by Teut2711** [reply]
I thought a lot and what I found out was  SerializerMethodField probably only works for ModelSerializer, otherwise I wasn't able to use it.
I could however solve my issue of adding field to validated_data by 

```python
 def validate(self, attrs):
        attrs = super().validate(attrs)
        attrs["HOL"] = self.get_HOL(attrs)
        attrs["PATST"] = self.get_PATST(attrs)
        return attrs

    def get_HOL(self, obj):
        return sum(
            filter(
                None,
                [
                    obj.get("FREEHOL,", 0),
                    obj.get("HOLLCK,", 0),
                    obj.get("HOLBLOCK,", 0),
                    obj.get("HOLPLD,", 0),
                    obj.get("HOLPLDLCK,", 0),
                    obj.get("HOLPLDUNC,", 0),
                    obj.get("HOLPLDLCKU,", 0),
                    obj.get("HOLREM,", 0),
                    obj.get("HOLREMLCK,", 0),
                    obj.get("HOLCMIDD,", 0),
                    obj.get("HOLCMPOOL,", 0),
                    obj.get("HOLSET,", 0),
                ],
            )
        )
```
 Few posts on SO, suggested to use an approach like
```python
PATST = serializers.IntegerField(source="get_PATST", read_only=True)
```
which include the computed fields data in the validated_data
